### ISSUE TITLE
Investigate replacing ZMQ with tuple queues

### ISSUE THREAD

**Issue by derekjn** [pipelinedb/pipelinedb#1848]
(Post extension refactor)

With the addition of parallel scans, PG obviously needed to establish a way for processes to pass tuples between each other, which is exactly what we use ZMQ for. They came up with a tuple queue for this purpose:

https://github.com/postgres/postgres/blob/113b0045e20d40f726a0a30e33214455e4f1385e/src/backend/executor/tqueue.c

The tuple queue use case is slightly different, simpler, and probably less performance critical than our current IPC path, but it's worth investigating if we can use these things for IPC and eliminate a dependency.

**Comment by usmanm** [comment]
I think tuple queue is for 1-1 IPC. We need a N-1 system. I think in general we need a better messaging layer. Something which offers better message level ordering guarantees. More of a Kafka-sequel interface.

**Comment by sangli00** [comment]
which one is the final choice?

**Comment by usmanm** [comment]
Nothing's decided yet. This probably won't be touched till after 1.0.

**Comment by sangli00** [comment]
I remember before used ```TupleBuffer``` build queue. sometimes,multi process is dead lock,Stream not get anything data. Queue  performance is very very important.

**Comment by usmanm** [comment]
Yes, that was an internal implementation we had. This one is built into PostgreSQL core and they use it for parallel aggregation. But we'll obviously not sacrifice performance or stability when making the decision.

**Comment by sangli00** [comment]
I looking forward to  PipelineDB extension release.

**Comment by usmanm** [comment]
Upcoming Redis Streams could be a potential candidate here as well. It's API is closely modeled around Kafka, but will likely me much faster.

**Comment by sangli00** [comment]
Redis Streams? This is new product?
